lib/src/cli/metrics/flutter_info.dart

```diff
@@ -0,0 +1,119 @@
+import 'utils.dart';
+
+part 'flutter_info.g.dart';
+
+FutureOr<T?> safe<T>(FutureOr<T> Function() f, {Function(Object e, StackTrace s)? onError}) async {
+  try {
+    return await f();
+  } catch (e, s) {
+    if (onError != null) {
+      onError(e, s);
+    }
+    return null;
+  }
+}
+
+class VersionConverter extends JsonConverter<Version, String> {
+  const VersionConverter();
+
+  @override
+  Version fromJson(String json) => Version.parse(json.takeUntil(' '));
+
+  @override
+  String toJson(Version object) => object.toString();
+}
+
+@JsonSerializable()
+class FlutterInfo {
+  @VersionConverter()
+  final Version frameworkVersion;
+  final String? channel;
+  final String? repositoryUrl;
+  final String? frameworkRevision;
+  final String? frameworkCommitDate;
+  final String? engineRevision;
+  @VersionConverter()
+  final Version dartSdkVersion;
+  final String? flutterRoot;
+
+  FlutterInfo(
+      {required this.frameworkVersion,
+      this.channel,
+      this.repositoryUrl,
+      this.frameworkRevision,
+      this.frameworkCommitDate,
+      this.engineRevision,
+      required this.dartSdkVersion,
+      this.flutterRoot});
+
+  factory FlutterInfo.fromJson(Map<String, dynamic> json) => _$FlutterInfoFromJson(json);
+
+  static Future<FlutterInfo?> get(Options options) async {
+    final pubspecPath = options.pubspecPath;
+    final pubspec = Pubspec.parse(await File(pubspecPath).readAsString());
+
+    const flutter = 'flutter';
+    final flutterDep = pubspec.dependencies.values.whereType<SdkDependency>().where((dep) => dep.sdk == flutter).firstOrNull;
+    if (flutterDep == null) {
+      return null; // no flutter dependency, so not a flutter project
+    }
+
+    // Read constraints, if any
+    var flutterVersionConstraints = flutterDep.version.intersect(pubspec.environment?[flutter] ?? VersionConstraint.any);
+
+    // Try to read actual version from version file in .dart_tools
+    final version = await safe(() async {
+      return Version.parse(await File(path.join(path.dirname(pubspecPath), '.dart_tool/version')).readAsString());
+    });
+
+    // Try to get full info by calling flutter executable
+    final info = await safe(() async {
+      final flutterRoot = options.flutterRoot;
+      final flutterPath = flutterRoot == null ? flutter : path.join(flutterRoot, 'bin', flutter);
+      final process = await Process.start(
+        flutterPath,
+        ['--version', '--machine'],
+      );
+      final infoJson = await process.stdout.transform(utf8.decoder).join();
+      return FlutterInfo.fromJson(json.decode(infoJson) as Map<String, dynamic>);
+    });
+
+    // Sanity check info
+    if (info != null && (version == null || version == info.frameworkVersion) && flutterVersionConstraints.allows(info.frameworkVersion)) {
+      return info; // the returned info match both the projects constraints and the flutter version used on last build
+    }
+
+    // Fallback to the version used on last build
```

Comment:
How is `info` having the version used in the last build here.

Also shouldn't the fallback be the flutter version constraint from the `pubspec`.

Reply:
version is updated when you run a flutter command, not when you run a dart one. Last build is perhaps a bit in-precise (`flutter pub get` will wright it fx). Let me tighten the comments

Reply:
ok I  opned it in editor to read the code better.

`version` here is not the last build. `.dart_tool/version` actually contains the current build. The one that is currently ongoing. 

So currently what is commented as fallback is actually a fallback if flutter_root was not found and we couldn't run the flutter executable. 

There should be one more fallback when neither flutter tool was run niether we found a `.dart_tool/version` then I think we need to report the flutter version from pubspec so we report at least what the project is using as minimum flutter version.

Reply:
Using the min of the flutter version. constraints as last fallback, before just sending 0.0.0